### PR TITLE
27 improve error handling

### PR DESCRIPTION
**Pull request by hetec**
Hopefully a way to solve our error issues:

- ApiError extends Error -> We can now use ApiError or one of the classes which inherit from ApiError if we want to raise a error (e.g. in reject functions)
- ApiError takes a specific ErrorType object which describes the error, so we don't need to check the error again e.g in the routes. We can simply use the information from the ApiError (like the message or the statusCode)
- All specific ErrorType are collected in a separate file
- All error messages are also collected in a separate file, so we can change the message in one place. And if we want to support I18n we can just switch the file. The errors must not change since the message keys remain the same
- You can still use parameters in the messages (See user not found for an example)

I really hope that makes sense, if not please tell me. This error stuff is a little bit terrible and complicated.

### CODE REVIEW

**Comment by flashback2k14** [comment]
looks really nice.

File: server/src/logic/services/auth.service.ts [hunk 1]
```diff
@@ -26,7 +25,19 @@ export class AuthService {
         .then((foundUser: SocoboUser) => this._cryptoUtils.comparePasswords(password, foundUser))
         .then((compareResult: ComparePwResult) => this._validateComparePasswords(compareResult.isPasswordMatch, compareResult.user))
         .then((user: SocoboUser) => resolve(this._createLoginResult(user)))
-        .catch((error: any) => reject(error));
+        .catch((error: any) => {
+          if (ErrorUtils.notFound(error)) {
+            let e = new DbError(ERRORS.USER_NOT_FOUND);
```
Comment (flashback2k14) [inline server/src/logic/services/auth.service.ts]: here is a .withArgs(usernameOrEmail) function call missing

File: server/src/routes/api/v1/logs/log.routes.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ export class LogRoute {
       (req: Request, res: Response, next: NextFunction) => {
         this._authValidator.checkValidToken(req)
           .then(() => next())
-          .catch((err: ApiError) => res.status(400).json(err));
+          .catch((err: any) => res.status(err.statusCode).json(err.forReponse()));
```
Comment (flashback2k14) [inline server/src/routes/api/v1/logs/log.routes.ts]: here is a typo in .forResponse()

File: server/src/error-messages.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+export const ERROR_MESSAGES: any = {
```
Comment (flashback2k14) [inline server/src/error-messages.ts]: i would move this into a i18n or messages folder

File: server/src/errors.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import { ErrorType } from "./models/index"
```
Comment (flashback2k14) [inline server/src/errors.ts]: i would move this into the models/errors folder

File: server/src/errors.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * First digit indicates the type:
+ * 0 => Generic errors
+ * 1 => User errors
+ * 2 => Auth errors
+ * 3 => Validation errors
+ */
+export class ERRORS {
```
Comment (flashback2k14) [inline server/src/errors.ts]: i would add a semicolon after each ErrorType

File: server/src/models/errors/api-error.ts [hunk 6]
```diff
@@ -1,29 +1,42 @@
...
+    let msg: string = ERROR_MESSAGES[messageKey];
+    let argCounter = 0
+    for (let arg of args) {
+      msg = msg.replace(`{${argCounter}}`, arg);
+      argCounter ++;
+    }
+    return msg;
+  }
+
+  forResponse = (): Object => {
```
Comment (flashback2k14) [inline server/src/models/errors/api-error.ts]: nice catch. only this function declaration works. 👍  the old forResponse(): Object {..} function declaration throws not found errros!

File: server/src/logic/utils/error.utils.ts [hunk 7]
```diff
@@ -7,4 +9,16 @@ export class ErrorUtils {
   static notFound (err: any): boolean {
     return err.code === errorCode.noData;
   }
+
+  static handleError(error: any, source: string, sourceMethod: string): ApiError{
```
Comment (flashback2k14) [inline server/src/logic/utils/error.utils.ts]: i like this error creation more than the long catch blocks. what did you think about use the builder pattern?

File: server/src/logic/services/auth.service.ts (hunk 1 above)
Comment (hetec) [inline server/src/logic/services/auth.service.ts]: Ok

File: server/src/errors.ts (hunk 5 above)
Comment (hetec) [inline server/src/errors.ts]: Ok

File: server/src/errors.ts (hunk 4 above)
Comment (hetec) [inline server/src/errors.ts]: Ok

File: server/src/error-messages.ts (hunk 3 above)
Comment (hetec) [inline server/src/error-messages.ts]: Ok

File: server/src/logic/utils/error.utils.ts (hunk 7 above)
Comment (hetec) [inline server/src/logic/utils/error.utils.ts]: Sounds good and  would End in a nice fluent style

**Comment by hetec** [comment]
I hope that's better. It is not exactly the builder pattern because I think this would be overkill for the error object but I think the fluent style of adding properties is much better than setting each property separatly

**Comment by flashback2k14** [comment]
Yes, this is much nicer. I like it.

**Comment by hetec** [comment]
What do you think about the new way of creating errors? Is this near to your idea of the builder pattern or had you something different in mind?

**Comment by flashback2k14** [comment]
I like it. We could merge this.

**Pull request merged**